rotkehlchen/chain/ethereum/modules/uniswap/v3/decoder.py

```diff
@@ -57,93 +69,240 @@ def _maybe_decode_v3_swap(  # pylint: disable=no-self-use
-                    # Those are assets returned due to a change in the swap price
-                    event.event_type = HistoryEventType.TRANSFER
-                    event.event_subtype = HistoryEventSubType.NONE
-                    event.counterparty = CPT_UNISWAP_V3
-                    event.notes = f'Refund of {event.balance.amount} {crypto_asset.symbol} in uniswap-v3 due to price change'  # noqa: E501
-
-        maybe_reshuffle_events(out_event=out_event, in_event=in_event)
+        if transaction.to_address not in UNISWAP_ROUTERS:
+            return decoded_events  # work only with the known routers for now
+
+        if len(decoded_events) < 3 or decoded_events[0].counterparty != CPT_GAS:
+            # The minimum is 3 events: gas, spend, receive.
+            return decoded_events
+
+        # Since we check that the to_address is a known router, we can be sure that the ETH
+        # transfer event (and other events) are parts of the swap.
+        if decoded_events[1].asset == A_ETH:
+            if decoded_events[1].event_type == HistoryEventType.SPEND:
+                # This is a swap from eth to token
+                swap_data = self._decode_eth_to_token_swap(decoded_events)
+            else:  # Receive
+                # This is a swap from token to eth
+                swap_data = self._decode_token_to_eth_swap(decoded_events)
+        else:
+            # Then this should be a swap from token to token
+            swap_data = self._decode_token_to_token_swap(decoded_events)
+
+        if swap_data is None or swap_data.from_asset is None or swap_data.to_asset is None:
+            self.msg_aggregator.add_warning(
+                f'Failed to decode a uniswap swap for transaction {transaction.tx_hash.hex()}. '
+                f'Please open an issue on our github or contact us in our discord server',
+            )
+            return decoded_events
+
+        try:
+            from_crypto_asset = swap_data.from_asset.resolve_to_crypto_asset()
+        except (UnknownAsset, WrongAssetType):
+            self.notify_user(event=decoded_events[1], counterparty=CPT_UNISWAP_V3)
+            return decoded_events
+
+        try:
+            to_crypto_asset = swap_data.to_asset.resolve_to_crypto_asset()
+        except (UnknownAsset, WrongAssetType):
+            self.notify_user(event=decoded_events[-1], counterparty=CPT_UNISWAP_V3)
+            return decoded_events
+
+        from_event = HistoryBaseEntry(
+            event_identifier=transaction.tx_hash,
+            sequence_index=decoded_events[1].sequence_index,
+            timestamp=decoded_events[1].timestamp,
+            location=Location.BLOCKCHAIN,
+            location_label=transaction.from_address,
+            asset=from_crypto_asset,
+            balance=Balance(amount=swap_data.from_amount),
+            notes=f'Swap {swap_data.from_amount} {from_crypto_asset.symbol} via uniswap auto router',  # noqa: E501
```

Comment:
Isn't this v3 specific?

```suggestion
            notes=f'Swap {swap_data.from_amount} {from_crypto_asset.symbol} via uniswap-v3 auto router',  # noqa: E501
```